x/concentrated-liquidity/incentives.go

```diff
@@ -49,3 +51,165 @@ func (k Keeper) getUptimeAccumulators(ctx sdk.Context, poolId uint64) ([]accum.A
+	}
+
+	// Get relevant pool-level values
+	poolIncentiveRecords := pool.GetPoolIncentives()
+	uptimeAccums, err := k.getUptimeAccumulators(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	// Since our base unit of time is nanoseconds, we divide with truncation by 10^9 (10e8) to get
+	// time elapsed in seconds
+	timeElapsedNanoSec := sdk.NewDec(int64(ctx.BlockTime().Sub(pool.GetLastLiquidityUpdate())))
+	timeElapsedSec := timeElapsedNanoSec.Quo(sdk.NewDec(10e8))
+
+	// If no time has elapsed, this function is a no-op
+	if timeElapsedSec.Equal(sdk.ZeroDec()) {
+		return nil
+	}
+
+	if timeElapsedSec.LT(sdk.ZeroDec()) {
+		return fmt.Errorf("Time elapsed cannot be negative.")
+	}
+
+	for uptimeIndex, uptimeAccum := range uptimeAccums {
+		// Get relevant uptime-level values
+		curUptimeDuration := types.SupportedUptimes[uptimeIndex]
+
+		// Qualifying liquidity is the amount of liquidity that satisfies uptime requirements
+		qualifyingLiquidity, err := uptimeAccum.GetTotalShares()
+		if err != nil {
+			return err
+		}
+
+		// If there is no qualifying liquidity for the current uptime accumulator, we leave it unchanged
+		if qualifyingLiquidity.LT(sdk.OneDec()) {
+			continue
+		}
+
+		incentivesToAddToCurAccum, updatedPoolRecords, err := calcAccruedIncentivesForAccum(ctx, curUptimeDuration, qualifyingLiquidity, timeElapsedSec, poolIncentiveRecords)
+		if err != nil {
+			return err
+		}
+
+		// Emit incentives to current uptime accumulator
+		uptimeAccum.AddToAccumulator(incentivesToAddToCurAccum)
+
+		// Update pool records (stored in state after loop)
+		poolIncentiveRecords = updatedPoolRecords
+	}
+
+	// Update pool incentive records and LastLiquidityUpdate time in state to reflect emitted incentives
+	pool.SetPoolIncentives(poolIncentiveRecords)
+	pool.SetLastLiquidityUpdate(ctx.BlockTime())
+	err = k.setPool(ctx, pool)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// nolint: unused
+// calcAccruedIncentivesForAccum calculates IncentivesPerLiquidity to be added to an accum
+// Returns the IncentivesPerLiquidity value and an updated list of IncentiveRecords that
+// reflect emitted incentives
+// Returns error if the qualifying liquidity/time elapsed are zero.
+func calcAccruedIncentivesForAccum(ctx sdk.Context, accumUptime time.Duration, qualifyingLiquidity sdk.Dec, timeElapsed sdk.Dec, poolIncentiveRecords []types.IncentiveRecord) (sdk.DecCoins, []types.IncentiveRecord, error) {
+	if !qualifyingLiquidity.IsPositive() || !timeElapsed.IsPositive() {
+		return sdk.DecCoins{}, []types.IncentiveRecord{}, fmt.Errorf("Qualifying liquidity and time elapsed must both be positive.")
+	}
+
+	incentivesToAddToCurAccum := sdk.NewDecCoins()
+	for incentiveIndex, incentiveRecord := range poolIncentiveRecords {
+		// We consider all incentives matching the current uptime that began emitting before the current blocktime
+		if incentiveRecord.StartTime.UTC().Before(ctx.BlockTime().UTC()) && incentiveRecord.MinUptime == accumUptime {
```

Comment:
My thinking was it would standardize across the different types of `time.Time` (e.g. `time.Now` has additional metadata that `time.Unix([current date])` does not have, even though both are technically the same type.

Reply:
I'm assuming that we initialize `incentiveRecord.StartTIme`  from `ctx.BlockTime()` 

They should already be standardized if my assumption is valid.